### ISSUE TITLE
Segfault when running backward on requires_grad=False

### ISSUE THREAD

**Issue by ezyang** [pytorch/pytorch#1611]
```
[ezyang@devgpu014.ash3 ~] python
Python 2.7.13 |Anaconda custom (64-bit)| (default, Dec 20 2016, 23:09:15) 
[GCC 4.4.7 20120313 (Red Hat 4.4.7-1)] on linux2
Type "help", "copyright", "credits" or "license" for more information.
Anaconda is brought to you by Continuum Analytics.
Please check out: http://continuum.io/thanks and https://anaconda.org
>>> from torch.autograd import Variable
>>> Variable(torch.randn(1), requires_grad=False).backward()
```

I'm investigating.

**Comment by apaszke** [comment]
[This](https://github.com/pytorch/pytorch/blob/master/torch/csrc/autograd/python_engine.cpp#L140) will be NULL in this case, but it will still be added to `roots`, causing the engine to segfault. It only happens for Variables that don't have a grad_fn (leaves) and have no grad_accumulator (requires_grad=False).

**Comment by colesbury** [comment]
Fixed in e1d257bc6d472ee297df1719bf344bae359dbeaa
